src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java

```
@@ -55,45 +75,88 @@
-    return null;
+  public Statement bind(Object identifier, Object value) {
+    requireNonNull(identifier);
+    if (identifier instanceof String) {
+      this.bindings.getLast().put((String)identifier, value);
+      return this;
+    }
+    throw new IllegalArgumentException("Only String identifiers are supported");
   }
 
   @Override
   public Statement bind(int i, Object o) {
-    return null;
+    throw new IllegalArgumentException("Only named parameters are supported");
   }
 
   @Override
-  public Statement bindNull(Object o, Class<?> type) {
-    return null;
+  public Statement bindNull(Object identifier, Class<?> type) {
+    return bind(identifier, null);
   }
 
   @Override
   public Statement bindNull(int i, Class<?> type) {
-    return null;
+    throw new IllegalArgumentException("Only named parameters are supported");
   }
 
   @Override
   public Publisher<? extends Result> execute() {
+    List<Struct> paramsStructs = new ArrayList<>();
+    for (Map<String, Object> bindingsBatch : this.bindings) {
+      Builder paramsStructBuilder = Struct.newBuilder();
+      Map<String, Type> types = this.types.isEmpty() ? new HashMap<>() : null;
+
+      for (Map.Entry<String, Object> binding : bindingsBatch.entrySet()) {
+        String paramName = binding.getKey();
+        Codec codec = this.resolvedCodecs.computeIfAbsent(paramName,
+            name -> this.codecs.getCodec(binding.getValue()));
+        paramsStructBuilder
+            .putFields(paramName, codec.encode(binding.getValue()));
+
+        if (this.types.isEmpty()) {
+          types.put(paramName,
+              Type.newBuilder().setCode(codec.getTypeCode()).build());
+        }
+
+      }
+
+      if (types != null) {
+        this.types = types;
+      }
+
+      paramsStructs.add(paramsStructBuilder.build());
+    }
+
+    Flux<Struct> structFlux = Flux.fromIterable(paramsStructs);
+
+    if (this.sql != null && this.sql.trim().toLowerCase().startsWith("select")) {
```

Could you add assertion for `sql` being non-null to the constructor? A statement with null sql would not be valid, so we should fail early on that.